app/src/org/commcare/engine/cases/CaseUtils.java

```diff
@@ -151,9 +126,80 @@ public static void purgeCases() throws InvalidCaseGraphException {
 
     }
 
+    private static Vector<String> getAllOwners() {
+        Vector<String> users = new Vector<>();
+        Vector<String> owners = new Vector<>();
+        SqlStorage<User> userStorage = CommCareApplication.instance().getUserStorage(User.STORAGE_KEY, User.class);
+        for (IStorageIterator<User> userIterator = userStorage.iterate(); userIterator.hasMore(); ) {
+            String id = userIterator.nextRecord().getUniqueId();
+            owners.addElement(id);
+            users.addElement(id);
+        }
+
+        //Now add all of the relevant groups
+        //TODO: Wow. This is.... kind of megasketch
+        for (String userId : users) {
+            DataInstance instance = CommCareUtil.loadFixture("user-groups", userId);
+            if (instance == null) {
+                continue;
+            }
+            EvaluationContext ec = new EvaluationContext(instance);
+            for (TreeReference ref : ec.expandReference(
+                    XPathReference.getPathExpr("/groups/group/@id").getReference())) {
+                AbstractTreeElement<AbstractTreeElement> idelement = ec.resolveReference(ref);
+                if (idelement.getValue() != null) {
+                    owners.addElement(idelement.getValue().uncast().getString());
+                }
+            }
+        }
+        return owners;
+    }
+
+    /**
+     * Get records ids for all related cases for the given set of cases including the original set of cases
+     *
+     * @param recordIds databse ids for the cases we want to find related cases
+     * @return database ids for all related cases for the given set of cases
+     */
+    public static Set<String> getRelatedCases(Set<String> recordIds) {
+        if (recordIds.isEmpty()) {
+            return recordIds;
+        }
+
+        SqlStorage<ACase> storage = CommCareApplication.instance().getUserStorage(ACase.STORAGE_KEY, ACase.class);
+        DAG<String, int[], String> fullCaseGraph = getFullCaseGraph(storage, new AndroidCaseIndexTable(),
+                new Vector<>());
+        Set<String> caseIds = getCaseIdsFromRecordsIds(storage, recordIds);
+        Set<String> relatedCaseIds = fullCaseGraph.getRelatedRecords(caseIds);
+        Set<String> relatedRecordIds = new HashSet<>();
+        for (String relatedCaseId : relatedCaseIds) {
+            String relatedRecordId = getRecordIdFromCaseId(storage, relatedCaseId);
+            relatedRecordIds.add(relatedRecordId);
+        }
+        return relatedRecordIds;
+    }
+
+    private static String getRecordIdFromCaseId(SqlStorage<ACase> storage, String caseId) {
+        Vector<Integer> result = storage.getIDsForValue(INDEX_CASE_ID, caseId);
+        if (result == null || result.isEmpty()) {
+            throw new IllegalStateException("No record found for case ID: " + caseId);
+        }
+        return String.valueOf(result.get(0));
```

Comment:
I'm a little confused a to why these ints are getting string casts in/out a bunch. I seems like this is could cause a fair amount of unnecessary memory churn (strings aren't exactly efficient), but would also communicate more clearly if these signatures stayed in integer as much as possible to prevent confusion between CaseID and RecordID.

Reply:
https://github.com/dimagi/commcare-android/pull/2955/commits/fdcabfc9a69461a2a53297d2cd1de09ed8cf7d22